### PR TITLE
[NFC] Change diagnostic text mentioning 'undeclared type'/'unresolved identifier'

### PR DESCRIPTION
**Pull request by varungandhi-apple**
TL;DR: I changed "use of undeclared type %0" -> "couldn't find definition of type %0", and "use of unresolved identifier" -> "couldn't find definition of"

## Undeclared type diagnostic

The current diagnostic seems unclear from a user-perspective as to what "undeclared type" means, especially if you have not written C/ObjC/C++ before, which _do_ have a distinct notion of declaring types in the surface language, as opposed to only being able to define types. It also potentially nudges the reader familiar with C/ObjC/C++ but new to Swift in an incorrect direction: "ok, this type is undeclared, can I declare it somehow and get rid of this error?" (The answer is you can't; you need to define it or import it.)

While the newer diagnostic is not _technically_ as accurate (e.g. you can have a declaration of a type in a swiftinterface without a definition), that's not really something people write in Swift code (unless this is a part of the language that I don't know about?). So I think the newer diagnostic matches the user model better.

For comparison, [Rust says](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=22859d114301d7b1ec5a4f7aa03fce60):

```
cannot find type `T` in this scope
```

which also seems like a reasonable alternative (although it mentions 'scope' which seems a bit unnecessary perhaps, even though it is technically correct).

Elm is a bit more verbose: (e.g. `type X = MakeX { x : Foo }`)

```
I cannot find a `Foo` type:
[..]
Hint: Read <https://elm-lang.org/0.19.1/imports> to see how `import`
declarations work in Elm.
```

## Unresolved identifier diagnostic

The basic reasoning is the same as before: "unresolved identifier" is compiler jargon which a user does not necessarily need to care about. Instead we can simply say that we could not find the definition. We could also do something like Rust where we say "cannot find $thing in scope".

## Misc Notes

1. I don't particularly care about the tense. If you think it should be `cannot` instead of `couldn't`, I'll change that, nbd.

2. I read through `docs/Diagnostics.md` and I don't think the recommendation of "phrase diagnostics as rules rather than reporting that the compiler failed to do something" is a good one. If we follow that recommendation, the diagnostic would look like:

    ```
    definition of $thing cannot be found in scope
    ```

    That seems overly aggressive IMO. Moreover, the point of the diagnostic here is not to describe a fundamental language rule (one about scoping and define-before-use), but to gently point out what went wrong. Of course, nothing's perfect, but I think it makes sense to _not_ follow that recommendation here.

### CODE REVIEW

**Comment by varungandhi-apple** [comment]
@swift-ci please test

**Comment by varungandhi-apple** [comment]
@swift-ci please test

File: include/swift/AST/DiagnosticsSema.def [hunk 1]
```diff
@@ -795,19 +795,19 @@ ERROR(unspaced_unary_operator,none,
...
       "did you mean to use '%2'?",
       (bool, StringRef, StringRef))
-ERROR(use_undeclared_type,none,
-      "use of undeclared type %0", (DeclNameRef))
-ERROR(use_undeclared_type_did_you_mean,none,
-      "use of undeclared type %0; did you mean to use '%1'?", (DeclNameRef, StringRef))
+ERROR(couldnt_find_defn_of_type,none,
+      "couldn't find definition of type %0", (DeclNameRef))
+ERROR(couldnt_find_defn_of_type_did_you_mean,none,
+      "couldn't find definition of type %0; did you mean to use '%1'?", (DeclNameRef, StringRef))
```
Comment (AnthonyLatsis) [inline include/swift/AST/DiagnosticsSema.def:810]: This is iffy of course, but I *do* think the full form (could not or cannot) would be more catchy and read better.

Comment (varungandhi-apple) [reply include/swift/AST/DiagnosticsSema.def:810]: I don't have a strong feeling about it either way. Since Holly has given a lot of thought to diagnostics and precise wording, I've left the decision up to her (noted in the PR text). 🙂

Comment (xedin) [reply include/swift/AST/DiagnosticsSema.def:810]: Indeed, as a rule we don’t use contractions in diagnostic text, I thought that Diagnostics.md already mentioned that but apparently not...

**Comment by varungandhi-apple** [comment]
Tracking rdar://problem/62375243.

File: include/swift/AST/DiagnosticsSema.def (hunk 1 above)
Comment (hborla) [reply include/swift/AST/DiagnosticsSema.def:810]: I agree with Varun that it doesn't make much of a difference, but I think we should use `cannot` for consistency.

**Comment by hborla** [comment]
> While the newer diagnostic is not technically as accurate (e.g. you can have a declaration of a type in a swiftinterface without a definition), that's not really something people write in Swift code (unless this is a part of the language that I don't know about?)

Could this be misleading in the context of protocol requirements? This is the one place I can think of where you're declaring things without providing a definition, and I'm wondering if that might cause confusion.

> I read through docs/Diagnostics.md and I don't think the recommendation of "phrase diagnostics as rules rather than reporting that the compiler failed to do something" is a good one.

I think at some point we should re-evaluate the rules laid out in Diagnostics.md. This particular rule does seem to encourage more clinical language, which is not very friendly to the user and often leads to these really jargony messages.

**Comment by varungandhi-apple** [comment]
> Could this be misleading in the context of protocol requirements? This is the one place I can think of where you're declaring things without providing a definition, and I'm wondering if that might cause confusion.

Like in a function signature `f<T : P>(_ t : T) -> ()`? Or something else?

**Comment by hborla** [comment]
I just mean in general. Say a user is trying to call a function that is a protocol requirement, and they make a typo. Now, they will get the error message `couldn't find definition of X`. This could be really misleading because the user might think the error has to do with the conformance to that protocol. The error message isn't telling them that `X` doesn't exist.

**Comment by varungandhi-apple** [comment]
I see what you mean, that makes sense. How about `cannot find type X in scope`/`cannot find X in scope`? That doesn't seem to have this problem. Any other ideas?

**Comment by varungandhi-apple** [comment]
[Note to self: The test failures were due to LLDB -- it also needs to be updated since it is matching on the diagnostics.]

**Comment by hborla** [comment]
I agree that dropping `definition` and being specific about the kind of entity (if we can) avoids this problem. I like `cannot find type X in scope`/`cannot find X in scope` (and I haven't come up with anything better). 

FWIW I don't think the word `scope` is jargon - this is a fundamental concept that programmers either will already know, or they will learn. This might be a good concept to explain in an educational note for this diagnostic.

**Comment by varungandhi-apple** [comment]
Updated code to use `cannot find X in scope`/`cannot find type X in scope`.

Matching LLDB PR: https://github.com/apple/llvm-project/pull/1128

@swift-ci please test

**Comment by varungandhi-apple** [comment]
apple/llvm-project#1128

@swift-ci please test

**Comment by varungandhi-apple** [comment]
apple/llvm-project#1128

@swift-ci please test

**Review by hborla — APPROVED** [review]
The more I look at the test cases, the more I like `cannot find [type/label/etc] X in scope`. The previous message always seemed to indicate that the thing doesn't exist, but `in scope` adds another angle – I've definitely hit this error before when the type I'm trying to use isn't in scope

**Comment by varungandhi-apple** [comment]
An LLDB test seems to be randomly failing. 😐 

apple/llvm-project#1128

@swift-ci please test
